pennylane/qnn/iqp.py

```diff
@@ -0,0 +1,387 @@
+        samples = csr_matrix(samples)
+        if generators_sp is None:
+            generators_sp = csr_matrix(generators)
+    else:
+        ops = csr_matrix(ops)
+        samples = csr_matrix(samples)
+
+    return samples, generators_sp, ops
+
+
+def _effective_params(gates, params):
+    par_transform = max(len(gate) for gate in gates) != 1
+    if par_transform:
+        effective_params = _par_transform(gates) @ params
+    else:
+        effective_params = params
+    return effective_params
+
+
+def _dense_samples(generators, samples, spin_sym):
+    samples_gates = 1 - 2 * ((samples @ generators.T) % 2)
+
+    samples_sum = []
+    samples_len = 0
+
+    if spin_sym:
+        samples_sum = samples.sum(axis=-1)
+        samples_len = samples.shape[0]
+
+    return samples_gates, samples_sum, samples_len
+
+
+def _dense_ops(generators, ops, spin_sym):
+    ops_gen = (ops @ generators.T) % 2
+
+    ops_sum = 0
+
+    if spin_sym:
+        ops_sum = ops.sum(axis=-1)
+
+    return ops_gen, ops_sum
+
+
+def _ini_spin_sym(ops_sum, samples_sum, samples_len, spin_sym):
+    if spin_sym:
+        try:
+            shape = (len(ops_sum), samples_len)
+        except TypeError:
+            shape = (samples_len,)
+
+        return 2 - jnp.repeat(ops_sum, samples_len).reshape(shape) % 2 - 2 * (samples_sum % 2)
+
+    return 1
+
+
+# pylint: disable=too-many-arguments
+def _op_expval_batch(
+    gates: list,
+    params: list,
+    n_qubits: int,
+    ops: list,
+    n_samples: int,
+    key: list,
+    spin_sym: bool = False,
+    sparse: bool = False,
+    indep_estimates: bool = False,
+) -> list:
+
+    if indep_estimates:
+        return _op_expval_indep(
+            gates=gates,
+            n_qubits=n_qubits,
+            params=params,
+            ops=ops,
+            n_samples=n_samples,
+            key=key,
+            sparse=sparse,
+            spin_sym=spin_sym,
+        )
+
+    samples = jax.random.randint(key, (n_samples, n_qubits), 0, 2)
+
+    generators = _generators(gates, n_qubits)
+    effective_params = _effective_params(gates, params)
+
+    generators_sp = None
+    if sparse:
+        generators_sp = _generators_sp(gates, n_qubits)
+
+    if sparse or isinstance(ops, csr_matrix):
+        samples, generators_sp, ops = _to_csr(generators, ops, samples, generators_sp)
+
+        samples_gates, samples_sum, samples_len = _sparse_samples(generators_sp, samples, spin_sym)
+        ops_gen, ops_sum = _sparse_ops(generators_sp, ops, spin_sym)
+
+    else:
+        samples_gates, samples_sum, samples_len = _dense_samples(generators, samples, spin_sym)
+        ops_gen, ops_sum = _dense_ops(generators, ops, spin_sym)
+
+    ini_spin_sym = _ini_spin_sym(ops_sum, samples_sum, samples_len, spin_sym)
+
+    par_ops_gates = 2 * effective_params * ops_gen
+    expvals = ini_spin_sym * jnp.cos(par_ops_gates @ samples_gates.T)
+
+    return expvals
+
+
+# pylint: disable=too-many-arguments
+def iqp_expval(
```

Comment:
Can this method be imported in the qnn/\_\_init\_\_.py so that it can be accessed as:

```
qml.qnn.iqp_expval
```